src/main/java/com/pss/ProjectileSimulator.java

```diff
@@ -24,24 +36,38 @@ public Projectile getProjectile() {
 
     public Vector3d updatePosition() {
         Vector3d newVelocity = getNewVelocity();
-        Vector3d newPosition = getNewPosition();
-
         _projectileHandler.getProjectile().setVelocity(newVelocity);
+
+        Vector3d newPosition = getNewPosition();
         _projectileHandler.getProjectile().setPosition(newPosition);
 
         return newPosition;
     }
 
     private Vector3d getNewVelocity() {
-        Vector3d currentVelocity = _projectileHandler.getProjectile().getVelocity();
-        if (currentVelocity == null) {
-            currentVelocity = new Vector3d(0, 0, 0);
-        }
+        Vector3d currentVelocity = new Vector3d(_projectileHandler.getProjectile().getVelocity());
+        Vector3d totalForce = new Vector3d(_projectileForceHandler.getProjectileForce());
+        totalForce.add(_projectileGravityHandler.getProjectileGravity());
+
+        Vector3d dragForce = new Vector3d(_projectileHandler.getProjectile().getVelocity());
```

Comment:
I think it makes more sense to just calculate what the resulting drag force will be, and add it instead of doing this subtracting and normalizing. You should be able to do simplify this alot by simply negating it in the handler.